barplotLinkageGroupCalls.func <- function(object, assemblyBED, by='lg', returnTable=FALSE)
{
	chr.table <- suppressWarnings(computeBarPlotMatrix(object, assemblyBED))
	chr.table <- chr.table[mixedsort(rownames(chr.table)),]

	chromoFrame <- melt(chr.table)
	colnames(chromoFrame) <- c('chr', 'LG', 'count')
	chromoFrame$LG <- factor(chromoFrame$LG, levels=mixedsort(levels(chromoFrame$LG)) )
	chromoFrame$chr <- factor(chromoFrame$chr, levels=mixedsort(levels(chromoFrame$chr)))

	#Plot by linkage group
	if(by=='lg')
	{	
		if(length(unique(chromoFrame$chr)) > 50){leg='none'}else{leg='right'}
		print(ggplot(chromoFrame, aes(LG, count))+
		geom_bar(stat="identity", aes(fill=chr), colour='black')+
		theme(axis.text.x = element_text(angle = 90, hjust = 1))+
		labs(x="Linkage Group", y="DNA Represented in Linkage Groups (Mb)")+
		theme(legend.position=leg)+
		ggtitle(paste("Barplot of ", length(unique(chromoFrame$chr)), " contigs clustered into ", length(unique(chromoFrame$LG)), " linkage groups",  sep="")))
	}
	
	#Alternately, plot by chromosome:
	if(by=='chr')
	{

		if(length(unique(chromoFrame$LG)) > 50){leg='none'}else{leg='right'}
		print(ggplot(chromoFrame, aes(chr, count))+
		geom_bar(stat="identity", aes(fill=LG), colour='black')+
		theme(axis.text.x = element_text(angle = 90, hjust = 1))+
		labs(x="Chromosome", y="DNA Represented in Chromosome (Mb)")+
		theme(legend.position=leg)+
		ggtitle(paste("Barplot of ", length(unique(chromoFrame$LG)), " linkage groups clustering into ", length(unique(chromoFrame$chr)), " chromosomes",  sep="")))
	}

	if(returnTable == TRUE){return(chr.table)}
}


####################################################################################################
#' Bar plot all linkage groups, with the true chromosomes of contigs coloured.
#' @param object LinkageGroupList, as generated by clusterContigs
#' @param assemblyBED table from a BED containing assembly information about the contigs, including length and chromosome
#' Note that the rownames of assemblyBED should be the contig names, as they are used in object, and the first column (chromosome name) will used to order by chromosome if 'chr' option used in by parameter. 
#' To use a bam file header, the product of makeChrTable(bamFile), or divideMyChr with or without the asBed option is suitable for input
#' @param by whether to plot by linkage group (if 'lg') or chromosomes ('chr')
#' @param returnTable TRUE to return chromosome length matrix.
#' Note to include legend, use legend=rownames(chr.table) for by='lg', and legend=colnames(chr.table) for by='chr'
#' 
#' @return a matrix of lengths of each chromosome (rows) in each linkage group (columns)
#' 
#' @aliases barplotLinkageGroupCalls barplotLinkageGroupCalls,LinkageGroupList,LinkageGroupList-method
#' @example inst/examples/barplotLinkageGroupCalls.R
#' @export
#' @import ggplot2
#' @importFrom reshape melt melt.matrix
#' @importFrom gtools mixedsort
####################################################################################################

setMethod('barplotLinkageGroupCalls',
          signature = signature(object = 'LinkageGroupList'),
          definition = barplotLinkageGroupCalls.func
)
